slack.js

```diff
@@ -0,0 +1,266 @@
+var _ = require("underscore");
+var league = require("./league.js");
+var fuzzy = require("./fuzzy_match.js");
+
+function StopControllerError () {}
+StopControllerError.prototype = new Error();
+
+var users = {
+    byName: {},
+    byId: {},
+    getId: function(name){
+        return this.byName[name].id;
+    },
+    getIdString: function(name){
+        return "<@"+this.getId(name)+">";
+    },
+    getByNameOrID: function(nameOrId) {
+        return this.byId[nameOrId] || this.byName[nameOrId];
+    }
+};
+var channels = {
+    byName: {},
+    byId: {},
+    getId: function(name){
+        return this.byName[name].id;
+    },
+    getIdString: function(name){
+        return "<#"+this.getId(name)+">";
+    }
+};
+
+function updateUser(bot){
+    // @ https://api.slack.com/methods/users.list
+    bot.api.users.list({}, function (err, response) {
+        if (err) {
+            throw new Error(err);
+        }
+
+        if (response.hasOwnProperty('members') && response.ok) {
+            var byName = {};
+            var byId = {};
+            var total = response.members.length;
+            for (var i = 0; i < total; i++) {
+                var member = response.members[i];
+                byName[member.name] = member;
+                byId[member.id] = member
+            }
+            users.byName = byName;
+            users.byId = byId;
+        }
+        console.log("info: got users");
+    });
+}
+
+function updateChannels(bot){
+    // @ https://api.slack.com/methods/channels.list
+    bot.api.channels.list({}, function (err, response) {
+        if (err) {
+            throw new Error(err);
+        }
+
+        if (response.hasOwnProperty('channels') && response.ok) {
+            var byName = {};
+            var byId = {};
+            var total = response.channels.length;
+            for (var i = 0; i < total; i++) {
+                var channel = response.channels[i];
+                byName[channel.name] = channel;
+                byId[channel.id] = channel;
+            }
+            channels.byName = byName;
+            channels.byId = byId;
+        }
+        console.log("info: got channels");
+    });
+
+}
+
+
+
+/* 
+   updates the user list
+   updates the channel lists
+   then repeats in new thread
+   
+   if it encounters an error it will exit the process with exit code 1
+*/
+var count = 0;
+function refresh(bot, delay, config) {
+    return criticalPath(Q.fcall(function(){
+        console.log("doing refresh " + count++);
+        bot.rtm.ping();
+        
+        updateUser(bot);
+        updateChannels(bot);
+        _.each(league.getAllLeagues(config), function(l) {
+            l.refresh();
+        });
+        setTimeout(function(){ 
+            refresh(bot, delay, config);
+        }, delay);
+    }));
+}
+
+function criticalPath(promise){
+    exceptionLogger(promise).catch(function(e) {
+        console.log("An exception was caught in a critical code-path. I am going down.");
```

Comment:
and this should be `console.error`

I believe this was just copy pasted from the old implementation so feel free to not touch it. just pointing it out.